### DISCUSSION TITLE
Converting python dataframes to Java ArrayLists

### DISCUSSION THREAD

**Discussion by nicoftrs** [the-anylogic-company/AnyLogic-Pypeline#62] [Q&A]
Hi,

So far I've been using pypeline to pass relatively simply outputs from Python scripts back to Java, typically as lists of floats that I converted to ArrayList<Double> like so:

ArrayList<Double> pythonOutput= pyCommunicator.runResults(ArrayList.class, "python_function(%s,%s,%s,%s)", param1, param2, param3, param4);

(here the output of python_function is a list of floats). 

However, I would now like to pass more complex data structures from python outputs, for example a dataframe with Dates as indexes and multiple columns of floats. When using the same command as shown above, but with python_function returning a dataframe with Date index, I get the following error: 

Cannot cast java.lang.String to java.util.ArrayList

I have therefore tried to set the output as a String and then apply the fromJson command: 

String pythonOutput= pyCommunicator.runResults(String.class, "python_function(%s,%s,%s,%s)", param1, param2, param3, param4);

ArrayList output = pyCommunicator.fromJson(pythonOutput , ArrayList.class);

However, in this case I get a Jackson error: 

com.fasterxml.jackson.core.JsonParseException: Unexpected character ('-' (code 45)): Expected space separating root-level values
 at [Source: (String)"2021-01-01 00:00:00    1.2
2021-01-01 01:00:00    1.2
2021-01-01 02:00:00    1.2

I have also tried to change the python output from dataframe to Json format using df.to_json(), but in this case I get the error 

Cannot deserialize value of type `java.util.ArrayList<java.lang.Object>` from Object value (token `JsonToken.START_OBJECT`)
 at [Source: (StringReader); line: 1, column: 1]

I'm probably mixing up different concepts in my attempt to solve the problem, I hope my explanation was more or less clear. 
Any hint or help would be much appreciated!

And of course, thank you for your very useful work so far, thinks makes Anylogic so much more powerful!

**Comment by t-wolfeadam** [comment]
It looks like you're on the right track! Your first attempt seems to fail because `python_function` is just returning a stringified version of the table, not valid JSON - so you're right to use `to_json`, but the default return from that function is an object, which is why it can't be converted to an ArrayList. 

Check out the [pandas docs for `to_json`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_json.html), particularly the examples section for samples of the different ways pandas can format the dataframe (and keep in mind which do/don't include the index). Here's my own example using something that looks like what you're doing:

![pandas to json samples](https://github.com/t-wolfeadam/AnyLogic-Pypeline/assets/40703071/ec34ba26-1d9b-443a-92f5-fd2e83d48b8e)

**Reply by nicoftrs** [reply]
Thanks for the reply! 

df.to_json(orient='table', date_format='iso') did indeed put it in the right format for me. However, I still had to tweek the json file output to get rid of all the metadata and only export the data values. 
I then implemented a little postprocessing loop in Anylogic to convert the date Strings into Date objects: 

String pythonOutput= pyCommunicator.runResults(String.class, "python_function(%s,%s,%s,%s)", param1, param2, param3, param4);
//Convert json string to list
List<HashMap<String, Object>> inputData = pyCommunicator.fromJson(pythonOutput, List.class);
//Postprocess each element in list
for (HashMap<String, Object> data : inputData) {
    //Convert date strings to Date format
    String sourceDateTime = (String) data.get("index");
    TimeZone utc = TimeZone.getTimeZone("UTC");
    SimpleDateFormat sourceFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
    sourceFormat.setTimeZone(utc);
    try{
        //Convert String to Date
    	Date convertedDate = sourceFormat.parse(sourceDateTime);
        //Extract corresponding double
        double value = (double)data.get("values");
        //Store in ArrayList of Pairs. 
	Pair<Date,Double> data_pair = new Pair<Date,Double>(convertedDate, value);
	dataValues.add(data_pair);
    } catch (ParseException e){
    	e.printStackTrace();
    }
}

Anyways, thanks for the help!

**Comment by elmar0103** [comment]
hi Nico, i am facing a similar problem, but do not manage to get it working. could you help me out?

**Reply by t-wolfeadam** [reply]
FYI no one can help you if you don't provide info on what your problem is and what your conditions are like (i.e., the model you're using, the code in the model, etc.). Based on your post on the other user's question, I suggest just opening a new issue/discussion thread.
